### ISSUE TITLE
Update the UI on incoming MIDI data

### ISSUE THREAD

**Issue by mkruselj** [samlletas/syxg50-control-panel#7]
For example, if we send the appropriate Bank Select+Program Change over a certain channel, we would see that program selected in Voice Catalog tab.

Likewise for any other parameter, if we receive the appropriate sysex or valid MIDI CC, that parameter should update on the UI.

**Comment by samlletas** [comment]
This is an interesting one :) 

The control panel uses a spinlock mechanism to prevent data-corruption due to concurrency (@gfx section runs in a different thread). In particular the @block section is locked only after receiving and passingthrough midi events (It wouldn't make sense to make midi messages wait), so one approach could be to push received cc/sysex messages into a queue and then update the plugin state after locking. For example:

```
@block
while (midirecv(offset,msg1,msg2,msg3)) (
   // Add events to queue.
   // Passthrough events.
);

spinlock.lock();

// Update plugin state based on queue.

spinlock.unlock();
```

We would probably need 2 queues, one for CCs and another one for sysex messages.

**Comment by kckcki** [comment]
@mkruselj  You read my mind squarely ! The plugin does Not seem to maintain the state of the currently selected or active Bank+Program. It should ideally be aware of the current "state" of the XG engine across all MIDI channels (XG parts).

**Comment by mkruselj** [comment]
Yep that would be ideal! 🙂

**Comment by samlletas** [comment]
> @mkruselj You read my mind squarely ! The plugin does Not seem to maintain the state of the currently selected or active Bank+Program. It should ideally be aware of the current "state" of the XG engine across all MIDI channels (XG parts).

The S-YXG50 synth as far as I know exposes no way to read it's current state, we can only send commands to it, but what we can do is update the plugin's UI based on incoming CCs and Sysex commands since those are passed through from the plugin to the S-YXG50 synth.

I plan to add this feature to v3, I'll probably do CCs first and leave Sysex for a later v3.x release since I believe those are lower priority (the only use-case I can think of for Sysex is playback of XG Midi files).

**Comment by mkruselj** [comment]
> but what we can do is update the plugin's UI based on incoming CCs and Sysex commands since those are passed through from the plugin to the S-YXG50 synth.

Yes, that is exactly what I proposed originally in the first post 🙂  We don't know the state of the plugin but we know the incoming data and we should update how we look based on it.

**Comment by samlletas** [comment]
> Yes, that is exactly what I proposed originally in the first post 🙂 We don't know the state of the plugin but we know the incoming data and we should update how we look based on it.

That's correct 🙂, I just wanted to clear up how this feature would work since the idea of being aware of the current XG engine state was mentioned in the comments.

**Comment by mkruselj** [comment]
Yeah one word was missing there, *incoming* 😛

**Comment by samlletas** [comment]
MIDI CC state updates are now supported, this feature will be included in the next version (v3).

Sysex state updates are yet to be implemented so I'll leave this GitHub issue open.

**Comment by kckcki** [comment]
> MIDI CC state updates are now supported, this feature will be included in the next version (v3).
> 
> Sysex state updates are yet to be implemented so I'll leave this GitHub issue open.

Wow ! That's wonderful to hear. This is a much-needed improvement Sam. Keep it up.

**Comment by mkruselj** [comment]
Absolutely brilliant, so so so useful!

Also, just as a sidenote, you did a really nice style for buttons. One suggestion I have is if you click on an already selected part of the button, it should flip it to the other state. So that no matter where you click on the button, it changes its value.

**Comment by samlletas** [comment]
> Also, just as a sidenote, you did a really nice style for buttons. One suggestion I have is if you click on an already selected part of the button, it should flip it to the other state. So that no matter where you click on the button, it changes its value.

Hmmm, the thing is that the new buttons are also used for some parameters with 3/4 options (Variation>Cross Delay, Variation>Amp Simulator) so they were initially intended more as an option selector rather than a checkbox.

One idea could be to increment the value on each click (like the part selector at the top-right) but that would mean that users won't be able to click on a specific value in parameters with more than 2 options. Another alternative could be to make it flip state only on parameters with 2 options.

**Comment by mkruselj** [comment]
Yeah that last thing - only flip state if 2 options are there. :)

**Comment by samlletas** [comment]
> Yeah that last thing - only flip state if 2 options are there. :)

Sounds good!
